### ISSUE TITLE
Error creating secret permission denied -- docker linux

### ISSUE THREAD

**Issue by mattlev** [izderadicka/audioserve#51]
Trying to create a docker container based on the izderadicka/audioserve:latest I continually run into a permission denied error for creating/reading the secret.  I have about 8 other containers that right to similar directories using the same PGID+PUID without error (everything being under a directory for the docker user each service having it's own "config" folder).  I've tried using the synology GUI updating all the info and command, I've tried sudo and nonsudo docker run commands via SSH, and tried creating the run-audioserve.sh file in the setup guide all to no avail.  Does the container setup accept PUID/PGID as environment inputs or does a hardcoded PUID have to be changed somewhere to the correct value for access to said directories?

**Comment by mattlev** [comment]
Trying to test with --no-authorization also gives permission denied when trying to read the /audiobooks directory

**Comment by izderadicka** [comment]
audioserve need a write permission to configuration directory  - usually it is mapped to external directory (volume)  `-v /path/for/audioserve-data:/home/audioserve/.audioserve ` .    Audioserve runs as user with UID 1000 GID 1000  (which is normally first regular user in Ubuntu, Debian ... -   so it works fine as they have same UID). 

So you have couple of options:
1) Asure that user with UID 1000 and GID 1000  has write access to that directory   (you can test chmod o+rwx  /path/for/audioserve-data - not ideal but you can verify that it is about permissions to that directory)
2)   Run container with different UID - see `--user` argument of docker run.   Remember - this user need to have read access to your audio collection and write permission to that configuration directory. 
3) Rebuild image - change user in Dockerfile - line 38:
```
RUN adduser -D -u 1000 audioserve &&\
```
Use here whatever UID is appropriate.

Couple of notes:

`sudo docker run`  - does not make difference - as container is started from daemon - important what is configured for container
 (via docker run arguments or Dockerfile)

You have two independent systems running - host and container - they can have different username (as per /etc/passwd) -   from file system perspective UID is important -   same UID = same user

**Comment by mattlev** [comment]
> 2\. Run container with different UID - see `--user` argument of docker run.

I tried this method but still get error denied (the same user id that is accessing the rest of the config folders and writing to them for my other docker containers)

Sounds like rebuilding off a different user id in the dockerfile will be what I would need to do, but presumably I would have to do that for every update?  I'll admit I don't know too much about dockerfiles and their commands, but is it possible for the dockerfile to know the commands inputted or are those only accessible in the image?  If you can point me to a source I'd be happy to take a stab at a pull request for a conditional that uses 1000 or a specific UID if one is given.

**Comment by izderadicka** [comment]
Yes you will need to rebuild with each update, which is bit more work (clone repo, run docker build).   The needed file for build is `Dockerfile` in root of repo. 

if you're running with different UID you need to realize that default data directory for audioserve will change (as by default it is in user's $HOME).

This is how I now tested audioserve running with different user:

```
docker run --rm --user 1003:1003 -p 3003:3000 -v /home/ivan/test_audiobooks/:/audiobooks -v /home/ivan/tmp/as_conf:/home/audioserve/.audioserve -e RUST_LOG=audioserve=debug -e AUDIOSERVE_SHARED_SECRET=nevim izderadicka/audioserve  /audiobooks --data-dir /home/audioserve/.audioserve
```

See `data-dir` argument to audioserve above. Mapping of data-dir is arbitrary - here using default location for user audioserve, but anything will do.  But on host side folder must exist and have proper permissions (write for user) -  because if it does not exist, then docker daemon creates this directory with root as owner.  
 
And the data-dir permissions (assure it's empty and does not contain files from previous attempts with incorrect owner): 

```
drwxrwxr-x  8 ivan ivan 4096 May  3 12:31 .
drwxr-xr-x 58 ivan ivan 4096 May  3 12:32 ..
drwxr-xr-x  2 1003 1003 4096 May  3 12:56 as_conf
```

Owner of directory is UID 1003 and audioserve is running as UID 1003.

**Comment by mattlev** [comment]
I've been setting the data directory to a static folder /volume/stuff (same as other other docker configs) so no relation to any user's home.

After looking into it, because another user is created during build I don't think there's a way to set it with a variable at runtime.  Anyone with non-default UID would have to build on their own 😢 
Looks like it's something done in the image(?) source that usually allows for customizable PUID/PGID (like a certain linux server group).

**Comment by izderadicka** [comment]
No, user in image (created by Dockerfile) can be overridden  by `docker run --user`.   Docker container is nothing else but process running under same kernel, but within much more restricted environment - cgroups, kernel namespaces, chroot ...,    so basic linux access rules just works: see ps related to my previous message:
```
$ ps -f -C audioserve
UID        PID  PPID  C STIME TTY          TIME CMD
1003      8267  8246  0 17:46 ?        00:00:00 ./audioserve /audiobooks --data-dir /home/audioserve/.audioserve
```

so audioserve is running as any other process - under UID 1003, so it has access to anything that this UID is entitled to. 
Related to "no relation to any user's home" - You have to understand there are two completely separate root file systems -

- one used by your host system - e.g file system you normally see
- files system of audioserve container  - which is independent, as container runs basically as chroot

So if you say that data should be in `/volume/stuff` it has not impact on where audioserve expects in container - the link between container folder an host folder is done via `-v` mapping - so what you need to do is (approximately):
- create subfolder  /volume/stuff/audioserve and assure it has write access for whatever  user audioserve is run with - <NEEDED_UID>
- run audioserve - docker run args ` -v /volume/stuff/audioserve:/home/audioserve/.audioserve --user <NEEDED_UID>`  and also audioserve with  an additional arg (audioserve args are after image name in docker cmd line)  `--data-dir /volume/stuff/audioserve`
